### ISSUE TITLE
brew update may not be updating all taps to latest HEAD

### ISSUE THREAD

**Issue by ilovezfs** [Homebrew/homebrew#49568]
This is from some ultimately successful troubleshooting on the IRC channel. The workaround that resolved the issue most simply was untap/tap, but obviously that's not a real fix. It may be that this is specific to the php tap. The user's log is below which shows that despite brew updating (even *before* the troubleshooting below), the head revision of homebrew/php wasn't updated to latest until an untap/tap. Note that memcached-2.2.0.tgz is the latest tarball not 2.1.0.

Following the update loop the user reported
>$ cd /usr/local/Library/Taps/homebrew/homebrew-php
>$ grep tgz Formula/php56-memcached.rb
> url 'http://pecl.php.net/get/memcached-2.1.0.tgz'

```
 /u/l/e/p/5.6 (master)> brew reinstall homebrew/php/php56-memcached
==> Reinstalling homebrew/php/php56-memcached
==> Downloading http://pecl.php.net/get/memcached-2.1.0.tgz
Already downloaded: /Library/Caches/Homebrew/php56-memcached-2.1.0.tgz
==> Downloading https://gist.githubusercontent.com/ablyler/6331007/raw/409ef282616859b2a7d19ba703c9b736576db16e/libmemcache-new-instance-
Already downloaded: /Library/Caches/Homebrew/php56-memcached--patch-bc958b0d50b7a260c13dc71451e6c35c08bc7c35.patch
==> Patching
==> Applying libmemcache-new-instance-api.patch
patching file memcached-2.1.0/ChangeLog
patching file memcached-2.1.0/README.markdown
patching file memcached-2.1.0/memcached-api.php
patching file memcached-2.1.0/memcached.ini
patching file memcached-2.1.0/php_memcached.c
patching file memcached-2.1.0/php_memcached.h
patching file memcached-2.1.0/php_memcached_session.c
Error: Unable to guess PHP branch for Formulary::FormulaNamespaceb4e23e325956ee5ea5e36c2e182eb5ce::Php56Memcached
 /u/l/e/p/5.6 (master)> for i in (seq 10)
                            brew update
                        end
Already up-to-date.
Already up-to-date.
Already up-to-date.
Already up-to-date.
Already up-to-date.
Already up-to-date.
Already up-to-date.
Already up-to-date.
Already up-to-date.
Already up-to-date.
 /u/l/e/p/5.6 (master)> brew config
HOMEBREW_VERSION: 0.9.5
ORIGIN: https://github.com/Homebrew/homebrew
HEAD: bd68469913811ea54247f3849dbbbabbbd9240a3
Last commit: 50 minutes ago
HOMEBREW_PREFIX: /usr/local
HOMEBREW_REPOSITORY: /usr/local
HOMEBREW_CELLAR: /usr/local/Cellar
HOMEBREW_BOTTLE_DOMAIN: https://homebrew.bintray.com
CPU: 8-core 64-bit haswell
OS X: 10.10.5-x86_64
Xcode: 7.2.1
CLT: 7.2.0.0.1.1447826929
Clang: 7.0 build 700
X11: 2.7.7 => /opt/X11
System Ruby: 2.0.0-p481
Perl: /usr/local/bin/perl => /usr/bin/perl
Python: /usr/local/bin/python => /usr/local/Cellar/python/2.7.11/Frameworks/Python.framework/Versions/2.7/bin/python2.7
Ruby: /usr/local/bin/ruby => /usr/local/Cellar/ruby/2.3.0/bin/ruby
Java: 1.8.0_40, 1.7.0_80, 1.7.0_79, 1.6.0_65-b14-466.1
 /u/l/e/p/5.6 (master)> cd /usr/local/Library/Taps/homebrew/homebrew-php
 /u/l/L/T/h/homebrew-php (master)> git status
On branch master
nothing to commit, working directory clean
 /u/l/L/T/h/homebrew-php (master)> git log | head -n 8
commit bd68469913811ea54247f3849dbbbabbbd9240a3
Author: Martin Afanasjew <martin@afanasjew.de>
Date:   Fri Feb 26 14:33:40 2016 +0100

    brew.sh: drop reference to removed 'update-ruby'
    
    This is a follow-up fix to a minor oversight in #49523.

 /u/l/L/T/h/homebrew-php (master)> brew untap homebrew/php
Untapping homebrew/php... (209 files, 2.1M)
Untapped 148 formulae
 /u/l/L/T/h/homebrew-php> cd 
 ~> brew tap homebrew/php
==> Tapping homebrew/php
Cloning into '/usr/local/Library/Taps/homebrew/homebrew-php'...
remote: Counting objects: 580, done.
remote: Compressing objects: 100% (368/368), done.
remote: Total 580 (delta 387), reused 242 (delta 210), pack-reused 0
Receiving objects: 100% (580/580), 239.90 KiB | 0 bytes/s, done.
Resolving deltas: 100% (387/387), done.
Checking connectivity... done.
Tapped 558 formulae (596 files, 1M)
 ~> cd /usr/local/Library/Taps/homebrew/homebrew-php
 /u/l/L/T/h/homebrew-php (master)> git log 
commit f21508926f29600bcc167b3fb9a4a663522fb6d0
Author: Andy Blyler <andy@blyler.cc>
Date:   Thu Feb 25 13:17:04 2016 -0500

    php5*-ssh2: update head to point to php5 branch
    ```

**Comment by ilovezfs** [comment]
@uniqmartin @xucheng FYI.

**Comment by UniqMartin** [comment]
The somewhat revealing part is:

```
 /u/l/L/T/h/homebrew-php (master)> git log | head -n 8
commit bd68469913811ea54247f3849dbbbabbbd9240a3
Author: Martin Afanasjew <martin@afanasjew.de>
Date:   Fri Feb 26 14:33:40 2016 +0100

    brew.sh: drop reference to removed 'update-ruby'

    This is a follow-up fix to a minor oversight in #49523.
```

This is a commit from the main Homebrew repository, so that indicates to me that the `Library/Taps/homebrew-homebrew-php` directory isn't even a Git repository at this stage and you're not viewing its history, because apparently there is none. Not surprisingly, since it is not a Git repository, `brew update` is unable to bring it up-to-date, no matter how often it is run.

The question is: How was that tap obtained? It apparently wasn't obtained via `brew tap` as that would have cloned a Git repository. That's what happened after `brew untap` and then `brew tap` again, so then it's up-to-date and the output of `git log | head -n 8` (shorter: `git log -1`) actually refers to the tap.

**Comment by ilovezfs** [comment]
@UniqMartin Good catch. Also it says it untapped 209 files / 148 formulae but tapped 596 files / 558 formulae, and , it hasn't been 2.1.0 since 12th March 2015. Totally weird.
